src/functions/function_utils.cpp

```diff
@@ -0,0 +1,121 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with MRCPP.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * For information on the complete list of contributors to MRCPP, see:
+ * <https://mrcpp.readthedocs.io/>
+ */
+
+#include "function_utils.h"
+#include "utils/details.h"
+
+#include "GaussExp.h"
+#include "Gaussian.h"
+#include "utils/Printer.h"
+
+#include <memory>
+
+namespace mrcpp {
+
+/** @brief Generates a GaussExp that is semi-periodic around a unit-cell
+ *
+ * @param[out] shared_ptr<GaussExp<D>> semi-periodic version of a Gaussian around a unit-cell
+ * @param[in] A Gaussian function that is to be made semi-periodic
+ * @param[in] The period of the unit cell
+ *
+ * The Gaussian function replicated in neighbooring cells, such that atleast
+ * 99.99367% if the integral can be captured within the unit-cell.
+ */
+template <int D>
+std::shared_ptr<GaussExp<D>> function_utils::make_gaussian_periodic(Gaussian<D> &inp,
+                                                                    const std::array<double, D> &period) {
+    auto gauss_exp = std::make_shared<GaussExp<D>>();
+    auto pos_vec = std::vector<Coord<D>>();
+
+    // Four standard deviation upp and down from the expectation value of the
+    // gaussian holds approximately 99.99367% of the integral.
+    auto four_std = 4.0 * inp.getMaximumStandardDiviation();
+
+    // This lambda function  calculates the number of neighbooring cells
+    // requred to keep atleast 99.99367% of the integral conserved in the
+    // unit-cell.
+    auto neighbooring_cells = [period, four_std](auto pos) {
+        auto needed_cells_vec = std::vector<int>();
+        for (auto i = 0; i < D; i++) {
+            auto upper_bound = pos[i] + four_std;
+            auto lower_bound = pos[i] - four_std;
+            // number of cells upp and down relative to the center of the Gaussian
+            needed_cells_vec.push_back(std::ceil(upper_bound / period[i]));
+        }
+
+        return *std::max_element(needed_cells_vec.begin(), needed_cells_vec.end());
+    };
+
+    // Finding starting position
+    auto startpos = inp.getPos();
+
+    for (auto d = 0; d < D; d++) {
+        startpos[d] = std::fmod(startpos[d], period[d]);
+        if (startpos[d] < 0) startpos[d] += period[d];
+    }
+
+    inp.setPos(startpos); // The starting position is set to complement build_grid
+                          // ensuring the projection of Gaussians with high exponents
+    auto nr_cells_upp_and_down = neighbooring_cells(startpos);
+    for (auto d = 0; d < D; d++) { startpos[d] -= nr_cells_upp_and_down * period[d]; }
+
+    auto tmp_pos = startpos;
+    for (auto x = 0; x < 2 * nr_cells_upp_and_down + 1; x++) {
+        if (D == 2 or D == 3) {
```

Comment:
This hurt my eyes! Can you try to rewrite using a recursive lambda function?
`auto` does not work in this case, so you'll have to spell it out, like
```
    std::function<int(int)> factorial;
    factorial = [&factorial](int i) {
        if (i == 0 or i == 1) {
            return 1;
        } else {
            return i*factorial(i-1);
        }
    };
```